### ISSUE TITLE
Crash on startup before initialization

### ISSUE THREAD

**Issue by yoshiyoshyosh** [Yellow-Dog-Man/Resonite-Issues#1230]
### Describe the bug?

after the `2024.1.23.613` hotfix, the game crashes on startup by not finding `ntdll.dll`

### To Reproduce

1. use linux native
2. open game
3. crash

### Expected behavior

the game doesn't crash

### Resonite Version Number

2024.1.23.613

### What Platforms does this occur on?

Linux

### Log Files

https://gist.github.com/yoshiyoshyosh/4cf01c986470ee1f79d4458a5c399074

### Additional Context

appears to once again be an issue of trying to load a native windows library on linux

### Reporters

yosh, @scary_guy_225

**Comment by ljoonal** [comment]
Seeing that fixing linux issues isn't a top priority, as this hasn't been fixed for multiple days now, I'll describe a workaround for those that do care.

```sh
dotnet tool install -g ilspycmd
export RESONITE_PATH="/home/lj/.steam/steam/steamapps/common/Resonite"
ilspycmd -d -p -o . -r "$RESONITE_PATH/Resonite_Data/Managed" "$RESONITE_PATH/Resonite_Data/Managed/System.Runtime.InteropServices.RuntimeInformation"
```

Then change the parts that try to use the ntdll to just return some value (preferrably one that makes it clear that it's modified so people don't waste time trying to diagnose based on your weird logs), and compile and replace the original file. I assume that sussy DLL files may not be too too welcome on this issue tracker or the officially managed discord server, thusly just a description of _how_ to work around the issue :sweat_smile:

**Comment by Scary-Guy** [comment]
> Seeing that fixing linux issues isn't a top priority, as this hasn't been fixed for multiple days now, here's a quick'n'dirty hacky workaround for those that do care.
> 
> ```shell
> dotnet tool install -g ilspycmd
> export RESONITE_PATH="/home/lj/.steam/steam/steamapps/common/Resonite"
> ilspycmd -d -p -o . -r "$RESONITE_PATH/Resonite_Data/Managed" "$RESONITE_PATH/Resonite_Data/Managed/System.Runtime.InteropServices.RuntimeInformation"
> ```
> 
> Then change the parts that try to use the ntdll to just return some value (preferrably one that makes it clear that it's modified so people don't waste time trying to diagnose based on your weird logs), `dotnet build`, and replace the original file. I assume that sussy DLL files may not be too too welcome on this issue tracker or the officially managed discord server, thusly just a description of _how_ to work around the issue 😅

My problem is I'm stuck with the suspect .dll files because I won't run .NET anything on my system.  I'll trust you over Micro$oft any day, but only because I have to.

**Comment by coolymike** [comment]
From what I was able to gather:
Between updates 2024.1.19.687 and now, the only things referencing `ntdll` (according to `grep -R "ntdll" ./`) are:
```
./Resonite_Data/Managed/System.Runtime.InteropServices.RuntimeInformation.dll
./Resonite_Data/Managed/FreeImageNET.dll
./Resonite_Data/Managed/Hardware.Info.dll
./Resonite_Data/Managed/FrooxEngine.dll
```

Of these, only the sha1sums of FrooxEngine differs.
```
76b0b6e65e38a90a4ecdb3f6dfe16d5a803081e9  ./Old/Resonite_Data/Managed/System.Runtime.InteropServices.RuntimeInformation.dll
76b0b6e65e38a90a4ecdb3f6dfe16d5a803081e9  ./New/Resonite_Data/Managed/System.Runtime.InteropServices.RuntimeInformation.dll
a8e8efa3e6c354c7942357d9562c1408d21ed8d2  ./Old/Resonite_Data/Managed/FreeImageNET.dll
a8e8efa3e6c354c7942357d9562c1408d21ed8d2  ./New/Resonite_Data/Managed/FreeImageNET.dll
5d350975e77e53adeb93ad10ba08e13834d64d80  ./Old/Resonite_Data/Managed/Hardware.Info.dll
5d350975e77e53adeb93ad10ba08e13834d64d80  ./New/Resonite_Data/Managed/Hardware.Info.dll
b318164bf51f714d07c63176495ae17c98282940  ./Old/Resonite_Data/Managed/FrooxEngine.dll
6fab6a3af23b93d0a943b7c27ab6c51729247142  ./New/Resonite_Data/Managed/FrooxEngine.dll
```

After having a quick look in a decompiled version of FrooxEngine, the following code in Engine.cs tries to load `ntdll.dll`:
```cs
[DllImport("ntdll.dll", CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Ansi, EntryPoint = "wine_get_version")]
private static extern string GetWineVersion();
```
This is in an attempt to detect Wine, and disable the desktop tab if under Wine. This code *does not seem to have changed* between a working older build, and the current, nonfunctional build.

The only other difference relating to this I can find is that `System.Runtime.InteropServices.RuntimeInformation` is now being used in `Engine.cs`:
```cs
UniLog.Log("Initializing App: " + CurrentVersion + $"\nOS Platform: {RuntimeInformation.OSDescription} ({RuntimeInformation.OSArchitecture})" + "\n" + systemInfo.PrintSystemInfo() + "\nMax GC Generation: " + GC.MaxGeneration + ", IsLittleEndian: " + BitConverter.IsLittleEndian + $"\nSystem.Numerics.Vectors HW accelerated: {Vector.IsHardwareAccelerated}, Vector<float>.Count: {Vector<float>.Count}" + $"\nBrotli native encoding/decoding supported: {DataTreeConverter.NativeBrotliSupported}");
```
This used to be the following in a working build of Resonite:
```cs
UniLog.Log("Initializing App: " + CurrentVersion + "\n" + systemInfo.PrintSystemInfo() + "\nMax GC Generation: " + GC.MaxGeneration + ", IsLittleEndian: " + BitConverter.IsLittleEndian + $"\nSystem.Numerics.Vectors HW accelerated: {Vector.IsHardwareAccelerated}, Vector<float>.Count: {Vector<float>.Count}" + $"\nBrotli native encoding/decoding supported: {DataTreeConverter.NativeBrotliSupported}");
```
That is most likely where the crash is coming from.

**Comment by Frooxius** [comment]
Yes, this seems to be coming from this package: https://www.nuget.org/packages/System.Runtime.InteropServices.RuntimeInformation/

The exception and the DLL's match.

It's a little bit ironic, since the goal of this package to detect the OS and check if we're running on Linux.

We'll have to remove the package, but problem is that some of our dependencies are using it too to detect the OS.

**Comment by coolymike** [comment]
Considering previously, Resonite did not crash due to this error, the dependencies mentioned are likely not using the functions in RuntimeInformation that cause the crash.

It might not be necessary to remove the entire package for now. Changing the log entry at initialization back to what it was should be enough to circumvent the crash.

It may also be an option to fork this library, only to add an explicit Windows check before trying to load ntdll.

**Comment by Frooxius** [comment]
Yes, it did not crash, but that doesn't mean this isn't concerning or something that we shouldn't look into.

If it's causing exception here and not detecting the system correctly, then it will do so for those dependencies and likely break some other behavior.

We can't really nicely fork this library, we'd have to swap to alternate Linux detection. There's ways to do it, but it will require us to modify those dependencies.

**Comment by Frooxius** [comment]
I've removed it from the main engine in 2024.1.30.731, this might resolve this, but I'm not sure what is the other impact of these changes right now.

**Comment by coolymike** [comment]
I was just able to launch Resonite native, only using the workaround for #69. I'm not sure if anything else broke due to the change.

**Comment by regalialong** [comment]
Can repro on Arch Linux that Native launches fine with only superpiss now, like Mike mentioned.

**Comment by phil-flip** [comment]
Can also reproduce on Native and Proton v9.0 and v8.0-5

**Comment by coolymike** [comment]
Running into another issue where System.Runtime.InteropServices is once again assuming it's on Windows, causing an exception when trying to use PDFiumSharp on Linux native.

**Comment by yoshiyoshyosh** [comment]
> Does this issue occur in the Windows build run via Proton as well, [@yoshiyoshyosh](https://github.com/yoshiyoshyosh)?

no. it was worked around while linux was not deprecated as well. though it might appear again if linux becomes un-deprecated again

**Comment by shiftyscales** [comment]
Thank you!
